ADAL/src/validation/ADAadAuthorityCache.m

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "ADAadAuthorityCache.h"
+
+#import "NSURL+ADExtensions.h"
+
+#include <pthread.h>
+
+@implementation ADAadAuthorityCacheRecord
+
+@end
+
+@implementation ADAadAuthorityCache
+
+- (id)init
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    
+    _map = [NSMutableDictionary new];
+    pthread_rwlock_init(&_rwLock, NULL);
+    
+    return self;
+}
+
+- (void)dealloc
+{
+    pthread_rwlock_destroy(&_rwLock);
+}
+
+- (void)processMetadata:(NSArray<NSDictionary *> *)metadata
+              authority:(NSURL *)authority
+                context:(id<ADRequestContext>)context
+{
+    [self getWriteLock];
+    
+    if (metadata.count == 0)
+    {
+        AD_LOG_INFO(@"No metadata returned from authority validation", context.correlationId, nil);
+    }
+    else
+    {
+        AD_LOG_INFO(@"Caching AAD Environements:", context.correlationId, nil);
+    }
+    
+    for (NSDictionary *environment in metadata)
+    {
+        __auto_type *record = [ADAadAuthorityCacheRecord new];
+        record.validated = YES;
+        record.networkHost = environment[@"preferred_network"];
+        record.cacheHost = environment[@"preferred_cache"];
+        
+        NSArray *aliases = environment[@"aliases"];
+        record.aliases = aliases;
+        
+        for (NSString *alias in aliases)
+        {
+            _map[alias] = record;
+        }
+        
+        AD_LOG_INFO(([NSString stringWithFormat:@"(%@, %@) : %@", record.networkHost, record.cacheHost, aliases]), context.correlationId, nil);
+    }
+    
+    // In case the authority we were looking for wasn't in the metadata
+    NSString *authorityHost = authority.adHostWithPortIfNecessary;
+    if (!_map[authorityHost])
```

Comment:
That's a big part of the motivation of wrapping the cache in this class. I'll have to add more checking specifically around cases where a port is for some odd reason specified, because I do think there is an issue in urlForPreferredHost where I probably shouldn't be checking url.host